storage/fragmentationCache/fragmentCache.cpp

```diff
@@ -0,0 +1,205 @@
+ * product may be distributed under the terms of the GNU General Public
+ * License (GPL) version 2 or later, in which case the provisions
+ * of the GPL apply INSTEAD OF those given above.
+ *
+ * This software is provided as is'', and any express or implied
+ * warranties, including, but not limited to, the implied warranties of
+ * merchantability and fitness for a particular purpose are disclaimed.
+ * In no event shall the company or contributors be liable for any
+ * direct, indirect, incidental, special, exemplary, or consequential
+ * damages (including, but not limited to, procurement of substitute
+ * goods or services; loss of use, data, or profits; or business
+ * interruption) however caused and on any theory of liability, whether
+ * in contract, strict liability, or tort (including negligence or
+ * otherwise) arising in any way out of the use of this software, even
+ * if advised of the possibility of such damage.
+ *
+ */
+
+#include "fragmentCache.hpp"
+#include "../xxhash.h"
+
+#include <cstring>
+
+namespace ipxp
+{
+
+// operators for working with timeval, the operators assume that:
+//   abs(time.usec) < USEC_IN_SEC && sign(time.sec) == sing(time.usec)
+
+static inline timeval operator-(const timeval &a, const timeval &b)
+{
+    // number of microseconds in second
+    constexpr decltype(a.tv_usec) USEC_IN_SEC = 1000000;
+
+    auto sec = a.tv_sec - b.tv_sec;
+    auto usec = a.tv_usec - b.tv_usec;
+
+    // ensure that abs(usec) < USEC_IN_SEC
+    if (usec < -USEC_IN_SEC) {
+        usec += USEC_IN_SEC;
+        --sec;
+    } else if (usec > USEC_IN_SEC) {
+        usec -= USEC_IN_SEC;
+        ++sec;
+    }
+
+    // ensure that sign(sec) == sign(usec)
+    if (sec > 0 && usec < 0) {
+        --sec;
+        usec += 1;
+    } else if (sec < 0 && usec > 0) {
+        ++sec;
+        usec -= 1;
+    }
+
+    return timeval{ .tv_sec = sec, .tv_usec = usec };
+}
+
+static constexpr bool operator>=(const timeval &a, const timeval &b)
+{
+    return a.tv_sec > b.tv_sec || (a.tv_sec == b.tv_sec && a.tv_usec >= b.tv_usec);
+}
+
+static constexpr bool operator==(const timeval &a, const timeval &b)
+{
+    return a.tv_sec == b.tv_sec && a.tv_usec == b.tv_usec;
+}
+
+FragmentCache::Key FragmentCache::Key::from_packet(Packet &pkt)
+{
+    return (Key) {
+        .ipv = pkt.ip_version,
+        .vlan_id = pkt.vlan_id,
+        .frag_id = pkt.frag_id,
+        .src_ip = pkt.src_ip,
+        .dst_ip = pkt.dst_ip
+    };
+}
+
+bool FragmentCache::Key::Equal::operator()(const Key &a, const Key &b) const
+{
+    if (a.ipv != b.ipv
+        || a.vlan_id != b.vlan_id
+        || a.frag_id != b.frag_id)
+    {
+        return false;
+    }
+
+    if (a.ipv == IP::v4) {
+        return a.src_ip.v4 == b.src_ip.v4 && a.dst_ip.v4 == b.dst_ip.v4;
+    }
+    return memcmp(a.src_ip.v6, b.src_ip.v6, sizeof(a.src_ip.v6)) == 0
+        && memcmp(a.dst_ip.v6, b.dst_ip.v6, sizeof(a.src_ip.v6)) == 0;
```

Comment:
Yes it can be done with a single memcmp, I did it this way because when the `ipv` is ipv4 you only need to compare the first four bytes of the `src_ip` and `dst_ip` because other bytes are 0.

Reply:
Yeah, I get it. However, from a CPU execution point of view, it would be much more optimal to use `memcmp()`:
- the compiler will be able to use fast vector units (i.e. SSE/AVX instructions)
- at the same time, most of the code branching will be removed, which in turn eliminates the impact of brach-predictors in the processor. This is because modern processors execute branches speculatively, and if there is a chance that the branching path of code execution changes frequently, this can have a significant impact on performance.

In other words, I would replace whole body of the function with a simple `memcmp()`.